feat(managers/custom): generic manager for json files (#32784)
renovatebot/renovate: docs/usage/configuration-options.md

- Proprietary file formats or conventions
- Popular file formats not yet supported as a manager by Renovate

> Currently we only have two custom managers.

Reviewer: ```suggestion
Currently we have two custom managers.
```

The `regex` manager which is based on using Regular Expression named capture groups.

> The `jsonata` manager which is based on using JSONata queries.

Author: Matching Engine sounds good.

You must capture/extract the following three fields _or_ configure its corresponding template (e.g. `depNameTemplate`) for these fields:

- `datasource`
- `depName` and / or `packageName`

> - `currentValue`

Author: I think the existing one is better.

> We recommend you use only _one_ of these methods, or you'll get confused.

Author: > For example, if you use a template for the first field, then use a template for the other fields too.

Sometimes one might have to combine both methods, so a recommendation should be fine here.

- Proprietary file formats or conventions
- Popular file formats not yet supported as a manager by Renovate

Renovate has two custom managers:

| Custom manager | Matching engine                                |
| -------------- | ---------------------------------------------- |
| `regex`        | Regular Expression, with named capture groups. |
| `jsonata`      | JSONata query.                                 |

You must capture/extract the following three fields _or_ configure its corresponding template (e.g. `depNameTemplate`) for these fields:

- `datasource`
- `depName` and / or `packageName`
- `currentValue`

We recommend you use only _one_ of these methods, or you'll get confused.

We also recommend that you also tell Renovate what `versioning` to use.

> If the `versioning` field is missing, then Renovate defaults to using `semver` versioning.

Reviewer 2: this isn't fully true. the datasource can set a different versioning (most data sources doing it). we should suggest this only if the datasource default versioning isn't suitable

> If the `versioning` field is missing, then Renovate defaults to using `semver` versioning.

Reviewer 3: Copying `@viceice`'s comment:

> this isn't fully true. the datasource can set a different versioning (most data sources doing it). we should suggest this only if the datasource default versioning isn't suitable

What do you think of this text?

```suggestion
Renovate defaults to `semver` versioning if _both_ condition are met:

- The `versioning` field is missing in the custom manager config
- The Renovate datasource does _not_ set its own default versioning
```

Reviewer 2: I need to check, but it probably uses `semver-coerced` as fallback 🤔

Reviewer 4: Yes, it is `semver-coerced`.

Reviewer 3: I updated my suggestion so it says `semver-coerced`. 😉

}
```

```json
{
  "customManagers": [
    {
      "customType": "jsonata",
      "matchStrings": [
        "packages.{ \"depName\": package, \"currentValue\": version }"
      ]
    }
  ]
}

> ```

Reviewer: Isn't this - and the example above it which you copy/pasted from - an incomplete `customManagers` example? e.g. missing `fileMatch`

}
```

> ```json

Reviewer 4: ```suggestion
```json title="Parsing a JSON file with a custom manager"
```

If `extractVersion` cannot be captured with a named capture group in `matchString` then it can be defined manually using this field.
It will be compiled using Handlebars and the regex `groups` result.

> ### matchQueries

Reviewer 2: can't we reuse `matchStrings`?

Author: Better to not have confusion

Reviewer: I'm not sure about this. I think I prefer `matchStrings` too, including because these two fields would be mutually exclusive

> Each `matchQueries` must be a valid regular expression(escaped).

Reviewer 2: regular expression? isn't it jsonata?

> Each `matchQueries` must be a valid regular expression, optionally with named capture groups.

Reviewer: ```suggestion
Each `matchQueries` must be a valid jsonata query (escaped).
```

And does it need or support more than one?

Author: Forgot to push this change 😅

Author: Yes, it supports more than one

### matchStrings

> Each `matchStrings` must be one of the two:
> 
> 1. a valid regular expression, optionally with named capture groups
> 2. a valid [JSONata](https://docs.jsonata.org/overview.html) query (escaped)

Reviewer: ```suggestion
Each `matchStrings` must be one of the two:

1. a valid regular expression, optionally with named capture groups (if using `customType=regex`)
2. a valid, escaped [JSONata](https://docs.jsonata.org/overview.html) query (if using customType=json)
```

If `extractVersion` cannot be captured with a named capture group in `matchString` then it can be defined manually using this field.
It will be compiled using Handlebars and the regex `groups` result.

### fileFormat

It specifies the syntax of the package file being managed by the custom JSONata manager.
This setting helps the system correctly parse and interpret the configuration file's contents.

Currently, only the `json` format is supported.

> ```json

Reviewer 4: ```suggestion
```json title="Parsing a JSON file with a custom manager"
```

> Currently, only the `json` format is supported. We intend to support more formats in the future.

Reviewer: ```suggestion
Currently, only the `json` format is supported.
```
